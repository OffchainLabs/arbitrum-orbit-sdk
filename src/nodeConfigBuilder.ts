import _set from 'lodash.set';

import { NodeConfig, NodeConfigOption } from './types/NodeConfig.generated';
import { CoreContracts } from './types/CoreContracts';
import { stringifyJson } from './nodeConfigBuilderUtils';
import { parentChainIsArbitrum } from './parentChainIsArbitrum';
import { nodeConfigBuilderDefaults } from './nodeConfigBuilderDefaults';
import { ParentChainId } from './types/ParentChain';
import { ChainConfig } from './types/ChainConfig';
import {
  NodeConfigChainInfoJson,
  NodeConfigDataAvailabilityRpcAggregatorBackendsJson,
} from './types/NodeConfig';

export type NodeConfigOptionKey = NodeConfigOption['key'];

export type NodeConfigOptionGetType<TKey extends NodeConfigOptionKey> = Extract<
  NodeConfigOption,
  { key: TKey }
>['type'];

export type NodeConfigBuilderEnableBatchPosterParams = {
  privateKey: string;
};

export type NodeConfigBuilderEnableStakerParams = {
  privateKey: string;
};

export type RpcAggregatorBackendsItem = {
  url: string;
  pubkey: string;
  signermask: number;
};

export type NodeConfigBuilderEnableDataAvailabilityServiceParams = {
  restAggregator: {
    urls: string[];
  };
  rpcAggregator: {
    assumedHonest?: undefined;
    backends: RpcAggregatorBackendsItem[];
  };
};

export class NodeConfigBuilder {
  /**
   * The underlying node config object being built.
   */
  private nodeConfig: NodeConfig;
  /**
   * Whether or not the builder was initialized with the necessary data.
   */
  private initialized: boolean;

  constructor(initialNodeConfig?: NodeConfig) {
    // it's necessary to do a spread operator here to make sure we create a copy of the object
    this.nodeConfig = { ...initialNodeConfig } ?? {};
    this.initialized = false;
  }

  private prepareChainInfoJson(params: InitializeThings): NodeConfigChainInfoJson {
    return [
      {
        'chain-id': params.chain.id,
        'chain-name': params.chain.name,
        'chain-config': params.chain.config,
        'parent-chain-id': params.parentChain.id,
        'parent-chain-is-arbitrum': parentChainIsArbitrum(params.parentChain.id),
        'rollup': {
          'bridge': params.parentChain.coreContracts.bridge,
          'inbox': params.parentChain.coreContracts.inbox,
          'sequencer-inbox': params.parentChain.coreContracts.sequencerInbox,
          'rollup': params.parentChain.coreContracts.rollup,
          'validator-utils': params.parentChain.coreContracts.validatorUtils,
          'validator-wallet-creator': params.parentChain.coreContracts.validatorWalletCreator,
          'deployed-at': params.parentChain.coreContracts.deployedAtBlockNumber,
        },
      },
    ];
  }

  private privateSet<TKey extends NodeConfigOptionKey>(
    key: TKey,
    value: NodeConfigOptionGetType<TKey>,
  ): NodeConfigBuilder {
    _set(this.nodeConfig, key, value);

    return this;
  }

  public initialize(params: InitializeThings): NodeConfigBuilder {
    const chainInfoJson = stringifyJson<NodeConfigChainInfoJson>(this.prepareChainInfoJson(params));

    this.privateSet('chain.name', params.chain.name);
    this.privateSet('chain.info-json', chainInfoJson);

    this.privateSet('parent-chain.id', params.parentChain.id);
    this.privateSet('parent-chain.connection.url', params.parentChain.rpcUrl);

    this.initialized = true;

    return this;
  }

  public enableSequencer(): NodeConfigBuilder {
    this.set('node.sequencer', true);

    this.set('node.delayed-sequencer.enable', true);
    this.set('node.delayed-sequencer.use-merge-finality', false);
    this.set('node.delayed-sequencer.finalize-distance', 1);

    this.set('node.dangerous.no-sequencer-coordinator', true);

    this.set('execution.forwarding-target', '');

    this.set('execution.sequencer.enable', true);
    this.set('execution.sequencer.max-tx-data-size', 85_000);
    this.set('execution.sequencer.max-block-speed', '250ms');

    this.set('execution.caching.archive', true);

    return this;
  }

  public enableBatchPoster(params: NodeConfigBuilderEnableBatchPosterParams): NodeConfigBuilder {
    this.set('node.batch-poster.enable', true);
    this.set('node.batch-poster.max-size', 90_000);
    this.set('node.batch-poster.parent-chain-wallet.private-key', params.privateKey);

    return this;
  }

  public enableStaker(params: NodeConfigBuilderEnableStakerParams): NodeConfigBuilder {
    this.set('node.staker.enable', true);
    this.set('node.staker.strategy', 'MakeNodes');
    this.set('node.staker.parent-chain-wallet.private-key', params.privateKey);

    return this;
  }

  public enableDataAvailabilityService(
    params: NodeConfigBuilderEnableDataAvailabilityServiceParams,
  ): NodeConfigBuilder {
    const backends = params.rpcAggregator.backends.map((backend, index) => ({
      ...backend,
      signermask: 1 << index, // 2^n
    }));

    const rpcAggregatorAssumedHonest = params.rpcAggregator.assumedHonest ?? 1;
    const rpcAggregatorBackendsJson =
      stringifyJson<NodeConfigDataAvailabilityRpcAggregatorBackendsJson>(backends);

    this.set('node.data-availability.enable', true);

    this.set('node.data-availability.rest-aggregator.enable', true);
    this.set('node.data-availability.rest-aggregator.urls', params.restAggregator.urls);

    this.set('node.data-availability.rpc-aggregator.enable', true);
    this.set('node.data-availability.rpc-aggregator.assumed-honest', rpcAggregatorAssumedHonest);
    this.set('node.data-availability.rpc-aggregator.backends', rpcAggregatorBackendsJson);

    return this;
  }

  public set<TKey extends NodeConfigOptionKey>(
    key: TKey,
    value: NodeConfigOptionGetType<TKey>,
  ): NodeConfigBuilder {
    if (!this.initialized) {
      throw new Error(`You must first call ".initialize()" on the builder`);
    }

    this.privateSet(key, value);

    return this;
  }

  build(): NodeConfig {
    return this.nodeConfig;
  }
}

type NodeConfigCoreContracts = Pick<
  CoreContracts,
  | 'bridge'
  | 'inbox'
  | 'sequencerInbox'
  | 'rollup'
  | 'validatorUtils'
  | 'validatorWalletCreator'
  | 'deployedAtBlockNumber'
>;

export type InitializeThings = {
  chain: {
    id: number;
    name: string;
    config: ChainConfig;
  };
  parentChain: {
    id: ParentChainId;
    rpcUrl: string;
    coreContracts: NodeConfigCoreContracts;
  };
};

export type CreateNodeConfigBuilderParams = {
  withDefaults?: boolean;
};

export function createNodeConfigBuilder(params?: CreateNodeConfigBuilderParams): NodeConfigBuilder {
  const withDefaults = params?.withDefaults ?? true;

  if (!withDefaults) {
    return new NodeConfigBuilder();
  }

  return new NodeConfigBuilder(nodeConfigBuilderDefaults);
}
