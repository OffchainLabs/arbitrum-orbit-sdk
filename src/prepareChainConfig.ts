import { ChainConfig, ChainConfigArbitrumParams } from './types/ChainConfig';

export const defaults = {
  homesteadBlock: 0,
  daoForkBlock: null,
  daoForkSupport: true,
  eip150Block: 0,
  eip150Hash: '0x0000000000000000000000000000000000000000000000000000000000000000',
  eip155Block: 0,
  eip158Block: 0,
  byzantiumBlock: 0,
  constantinopleBlock: 0,
  petersburgBlock: 0,
  istanbulBlock: 0,
  muirGlacierBlock: 0,
  berlinBlock: 0,
  londonBlock: 0,
  clique: {
    period: 0,
    epoch: 0,
  },
  arbitrum: {
    EnableArbOS: true,
    AllowDebugPrecompiles: false,
    DataAvailabilityCommittee: false,
    InitialArbOSVersion: 20,
    GenesisBlockNum: 0,
    MaxCodeSize: 24_576,
    MaxInitCodeSize: 49_152,
  },
};

export type PrepareChainConfigParams = Pick<ChainConfig, 'chainId'> &
  Partial<Omit<ChainConfig, 'chainId' | 'arbitrum'>> & {
    arbitrum: Pick<ChainConfigArbitrumParams, 'InitialChainOwner'> &
      Partial<Omit<ChainConfigArbitrumParams, 'InitialChainOwner'>>;
  };

export function prepareChainConfig(params: PrepareChainConfigParams): ChainConfig {
  return {
    ...defaults,
    ...params,
    clique: { ...defaults.clique, ...params.clique },
    arbitrum: { ...defaults.arbitrum, ...params.arbitrum },
  };
}
