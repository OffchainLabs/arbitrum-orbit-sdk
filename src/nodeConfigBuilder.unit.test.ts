import { it, expect } from 'vitest';

import { createNodeConfigBuilder, InitializeThings } from './nodeConfigBuilder';
import { prepareChainConfig } from './prepareChainConfig';

const initializeParams: InitializeThings = {
  chain: {
    id: 123_456,
    name: 'my-l3-chain',
    config: prepareChainConfig({
      chainId: 123_456,
      arbitrum: { InitialChainOwner: '0x1000000000000000000000000000000000000000' },
    }),
  },
  parentChain: {
    id: 421_614,
    rpcUrl: 'https://sepolia-rollup.arbitrum.io/rpc',
    coreContracts: {
      bridge: '0x0000000000000000000000000000000000000001',
      inbox: '0x0000000000000000000000000000000000000002',
      sequencerInbox: '0x0000000000000000000000000000000000000003',
      rollup: '0x0000000000000000000000000000000000000004',
      validatorUtils: '0x0000000000000000000000000000000000000005',
      validatorWalletCreator: '0x0000000000000000000000000000000000000006',
      deployedAtBlockNumber: 456_789,
    },
  },
};

it('creates node config', () => {
  const nodeConfig = createNodeConfigBuilder()
    //
    .initialize(initializeParams)
    .build();

  expect(nodeConfig).toMatchSnapshot();
});

it('creates node config without defaults', () => {
  const nodeConfig = createNodeConfigBuilder({ withDefaults: false })
    .initialize(initializeParams)
    .build();

  expect(nodeConfig).toMatchSnapshot();
});

it('creates node config with data availability service', () => {
  const nodeConfig = createNodeConfigBuilder()
    .initialize(initializeParams)
    .enableDataAvailabilityService({
      restAggregator: {
        urls: ['http://localhost:9877'],
      },
      rpcAggregator: {
        backends: [
          {
            url: 'http://localhost:9876',
            pubkey:
              'YAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==',
            signermask: 1,
          },
        ],
      },
    })
    .build();

  expect(nodeConfig).toMatchSnapshot();
});

it('creates node config with sequencer, batch poster and staker', () => {
  const nodeConfig = createNodeConfigBuilder()
    .initialize(initializeParams)
    .enableSequencer()
    .enableBatchPoster({ privateKey: 'BATCH_POSTER_PRIVATE_KEY' })
    .enableStaker({ privateKey: 'STAKER_PRIVATE_KEY' })
    .build();

  expect(nodeConfig).toMatchSnapshot();
});
