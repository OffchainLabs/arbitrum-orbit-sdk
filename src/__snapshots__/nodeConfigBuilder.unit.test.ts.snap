// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`creates node config 1`] = `
{
  "chain": {
    "info-json": "[{\\"chain-id\\":123456,\\"chain-name\\":\\"my-l3-chain\\",\\"chain-config\\":{\\"homesteadBlock\\":0,\\"daoForkBlock\\":null,\\"daoForkSupport\\":true,\\"eip150Block\\":0,\\"eip150Hash\\":\\"0x0000000000000000000000000000000000000000000000000000000000000000\\",\\"eip155Block\\":0,\\"eip158Block\\":0,\\"byzantiumBlock\\":0,\\"constantinopleBlock\\":0,\\"petersburgBlock\\":0,\\"istanbulBlock\\":0,\\"muirGlacierBlock\\":0,\\"berlinBlock\\":0,\\"londonBlock\\":0,\\"clique\\":{\\"period\\":0,\\"epoch\\":0},\\"arbitrum\\":{\\"EnableArbOS\\":true,\\"AllowDebugPrecompiles\\":false,\\"DataAvailabilityCommittee\\":false,\\"InitialArbOSVersion\\":11,\\"GenesisBlockNum\\":0,\\"MaxCodeSize\\":24576,\\"MaxInitCodeSize\\":49152,\\"InitialChainOwner\\":\\"0x1000000000000000000000000000000000000000\\"},\\"chainId\\":123456},\\"parent-chain-id\\":421614,\\"parent-chain-is-arbitrum\\":true,\\"rollup\\":{\\"bridge\\":\\"0x0000000000000000000000000000000000000001\\",\\"inbox\\":\\"0x0000000000000000000000000000000000000002\\",\\"sequencer-inbox\\":\\"0x0000000000000000000000000000000000000003\\",\\"rollup\\":\\"0x0000000000000000000000000000000000000004\\",\\"validator-utils\\":\\"0x0000000000000000000000000000000000000005\\",\\"validator-wallet-creator\\":\\"0x0000000000000000000000000000000000000006\\",\\"deployed-at\\":456789}}]",
    "name": "my-l3-chain",
  },
  "http": {
    "addr": "0.0.0.0",
    "api": [
      "eth",
      "net",
      "web3",
      "arb",
      "debug",
    ],
    "corsdomain": [
      "*",
    ],
    "port": 8449,
    "vhosts": [
      "*",
    ],
  },
  "parent-chain": {
    "connection": {
      "url": "https://sepolia-rollup.arbitrum.io/rpc",
    },
    "id": 421614,
  },
}
`;

exports[`creates node config with data availability service 1`] = `
{
  "chain": {
    "info-json": "[{\\"chain-id\\":123456,\\"chain-name\\":\\"my-l3-chain\\",\\"chain-config\\":{\\"homesteadBlock\\":0,\\"daoForkBlock\\":null,\\"daoForkSupport\\":true,\\"eip150Block\\":0,\\"eip150Hash\\":\\"0x0000000000000000000000000000000000000000000000000000000000000000\\",\\"eip155Block\\":0,\\"eip158Block\\":0,\\"byzantiumBlock\\":0,\\"constantinopleBlock\\":0,\\"petersburgBlock\\":0,\\"istanbulBlock\\":0,\\"muirGlacierBlock\\":0,\\"berlinBlock\\":0,\\"londonBlock\\":0,\\"clique\\":{\\"period\\":0,\\"epoch\\":0},\\"arbitrum\\":{\\"EnableArbOS\\":true,\\"AllowDebugPrecompiles\\":false,\\"DataAvailabilityCommittee\\":false,\\"InitialArbOSVersion\\":11,\\"GenesisBlockNum\\":0,\\"MaxCodeSize\\":24576,\\"MaxInitCodeSize\\":49152,\\"InitialChainOwner\\":\\"0x1000000000000000000000000000000000000000\\"},\\"chainId\\":123456},\\"parent-chain-id\\":421614,\\"parent-chain-is-arbitrum\\":true,\\"rollup\\":{\\"bridge\\":\\"0x0000000000000000000000000000000000000001\\",\\"inbox\\":\\"0x0000000000000000000000000000000000000002\\",\\"sequencer-inbox\\":\\"0x0000000000000000000000000000000000000003\\",\\"rollup\\":\\"0x0000000000000000000000000000000000000004\\",\\"validator-utils\\":\\"0x0000000000000000000000000000000000000005\\",\\"validator-wallet-creator\\":\\"0x0000000000000000000000000000000000000006\\",\\"deployed-at\\":456789}}]",
    "name": "my-l3-chain",
  },
  "http": {
    "addr": "0.0.0.0",
    "api": [
      "eth",
      "net",
      "web3",
      "arb",
      "debug",
    ],
    "corsdomain": [
      "*",
    ],
    "port": 8449,
    "vhosts": [
      "*",
    ],
  },
  "node": {
    "data-availability": {
      "enable": true,
      "rest-aggregator": {
        "enable": true,
        "urls": [
          "http://localhost:9877",
        ],
      },
      "rpc-aggregator": {
        "assumed-honest": 1,
        "backends": "[{\\"url\\":\\"http://localhost:9876\\",\\"pubkey\\":\\"YAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==\\",\\"signermask\\":1}]",
        "enable": true,
      },
    },
  },
  "parent-chain": {
    "connection": {
      "url": "https://sepolia-rollup.arbitrum.io/rpc",
    },
    "id": 421614,
  },
}
`;

exports[`creates node config with sequencer, batch poster and staker 1`] = `
{
  "chain": {
    "info-json": "[{\\"chain-id\\":123456,\\"chain-name\\":\\"my-l3-chain\\",\\"chain-config\\":{\\"homesteadBlock\\":0,\\"daoForkBlock\\":null,\\"daoForkSupport\\":true,\\"eip150Block\\":0,\\"eip150Hash\\":\\"0x0000000000000000000000000000000000000000000000000000000000000000\\",\\"eip155Block\\":0,\\"eip158Block\\":0,\\"byzantiumBlock\\":0,\\"constantinopleBlock\\":0,\\"petersburgBlock\\":0,\\"istanbulBlock\\":0,\\"muirGlacierBlock\\":0,\\"berlinBlock\\":0,\\"londonBlock\\":0,\\"clique\\":{\\"period\\":0,\\"epoch\\":0},\\"arbitrum\\":{\\"EnableArbOS\\":true,\\"AllowDebugPrecompiles\\":false,\\"DataAvailabilityCommittee\\":false,\\"InitialArbOSVersion\\":11,\\"GenesisBlockNum\\":0,\\"MaxCodeSize\\":24576,\\"MaxInitCodeSize\\":49152,\\"InitialChainOwner\\":\\"0x1000000000000000000000000000000000000000\\"},\\"chainId\\":123456},\\"parent-chain-id\\":421614,\\"parent-chain-is-arbitrum\\":true,\\"rollup\\":{\\"bridge\\":\\"0x0000000000000000000000000000000000000001\\",\\"inbox\\":\\"0x0000000000000000000000000000000000000002\\",\\"sequencer-inbox\\":\\"0x0000000000000000000000000000000000000003\\",\\"rollup\\":\\"0x0000000000000000000000000000000000000004\\",\\"validator-utils\\":\\"0x0000000000000000000000000000000000000005\\",\\"validator-wallet-creator\\":\\"0x0000000000000000000000000000000000000006\\",\\"deployed-at\\":456789}}]",
    "name": "my-l3-chain",
  },
  "execution": {
    "caching": {
      "archive": true,
    },
    "forwarding-target": "",
    "sequencer": {
      "enable": true,
      "max-block-speed": "250ms",
      "max-tx-data-size": 85000,
    },
  },
  "http": {
    "addr": "0.0.0.0",
    "api": [
      "eth",
      "net",
      "web3",
      "arb",
      "debug",
    ],
    "corsdomain": [
      "*",
    ],
    "port": 8449,
    "vhosts": [
      "*",
    ],
  },
  "node": {
    "batch-poster": {
      "enable": true,
      "max-size": 90000,
      "parent-chain-wallet": {
        "private-key": "BATCH_POSTER_PRIVATE_KEY",
      },
    },
    "dangerous": {
      "no-sequencer-coordinator": true,
    },
    "delayed-sequencer": {
      "enable": true,
      "finalize-distance": 1,
      "use-merge-finality": false,
    },
    "sequencer": true,
    "staker": {
      "enable": true,
      "parent-chain-wallet": {
        "private-key": "STAKER_PRIVATE_KEY",
      },
      "strategy": "MakeNodes",
    },
  },
  "parent-chain": {
    "connection": {
      "url": "https://sepolia-rollup.arbitrum.io/rpc",
    },
    "id": 421614,
  },
}
`;

exports[`creates node config without defaults 1`] = `
{
  "chain": {
    "info-json": "[{\\"chain-id\\":123456,\\"chain-name\\":\\"my-l3-chain\\",\\"chain-config\\":{\\"homesteadBlock\\":0,\\"daoForkBlock\\":null,\\"daoForkSupport\\":true,\\"eip150Block\\":0,\\"eip150Hash\\":\\"0x0000000000000000000000000000000000000000000000000000000000000000\\",\\"eip155Block\\":0,\\"eip158Block\\":0,\\"byzantiumBlock\\":0,\\"constantinopleBlock\\":0,\\"petersburgBlock\\":0,\\"istanbulBlock\\":0,\\"muirGlacierBlock\\":0,\\"berlinBlock\\":0,\\"londonBlock\\":0,\\"clique\\":{\\"period\\":0,\\"epoch\\":0},\\"arbitrum\\":{\\"EnableArbOS\\":true,\\"AllowDebugPrecompiles\\":false,\\"DataAvailabilityCommittee\\":false,\\"InitialArbOSVersion\\":11,\\"GenesisBlockNum\\":0,\\"MaxCodeSize\\":24576,\\"MaxInitCodeSize\\":49152,\\"InitialChainOwner\\":\\"0x1000000000000000000000000000000000000000\\"},\\"chainId\\":123456},\\"parent-chain-id\\":421614,\\"parent-chain-is-arbitrum\\":true,\\"rollup\\":{\\"bridge\\":\\"0x0000000000000000000000000000000000000001\\",\\"inbox\\":\\"0x0000000000000000000000000000000000000002\\",\\"sequencer-inbox\\":\\"0x0000000000000000000000000000000000000003\\",\\"rollup\\":\\"0x0000000000000000000000000000000000000004\\",\\"validator-utils\\":\\"0x0000000000000000000000000000000000000005\\",\\"validator-wallet-creator\\":\\"0x0000000000000000000000000000000000000006\\",\\"deployed-at\\":456789}}]",
    "name": "my-l3-chain",
  },
  "parent-chain": {
    "connection": {
      "url": "https://sepolia-rollup.arbitrum.io/rpc",
    },
    "id": 421614,
  },
}
`;
